File: apps/api/app/modules/provider_polling/adapters/mock.py
from datetime import UTC, datetime
from decimal import Decimal

from app.modules.candles.timeframes import timeframe_duration
from app.modules.provider_polling.adapters.base import (
    ProviderPollingAdapter,
    ProviderPollingFetchRequest,
    ProviderPollingResult,
)
from app.modules.provider_polling.schemas import ProviderCandle


class MockPollingProviderAdapter(ProviderPollingAdapter):
    provider_key = "mock_polling"

    async def fetch_candles(
        self,
        request: ProviderPollingFetchRequest,
    ) -> ProviderPollingResult:
        candles: list[ProviderCandle] = []
        current_timestamp = request.start_time or default_start_time()
        duration = timeframe_duration(request.timeframe)
        for index in range(request.limit):
            if request.end_time is not None and current_timestamp > request.end_time:
                break
            open_price = Decimal("65000.00") + Decimal(index)
            close_price = open_price + Decimal("5.00")
            candles.append(
                ProviderCandle(
                    provider_symbol=request.provider_symbol,
                    timeframe=request.timeframe,
                    timestamp=current_timestamp,
                    open=open_price,
                    high=close_price + Decimal("10.00"),
                    low=open_price - Decimal("10.00"),
                    close=close_price,
                    volume=Decimal("10.00") + Decimal(index),
                    is_final=True,
                    raw_item_json={
                        "provider": self.provider_key,
                        "sequence": index + 1,
                        "providerSymbol": request.provider_symbol,
                    },
                )
            )
            current_timestamp += duration
        return ProviderPollingResult(
            candles=candles,
            provider_metadata={
                "provider": self.provider_key,
                "requested_url": None,
                "generated": True,
            },
            warnings=[],
            errors=[],
        )


def default_start_time() -> datetime:
    return datetime(2026, 4, 29, 10, 0, tzinfo=UTC)

File: apps/go/market-worker/internal/providers/mock.go
package providers

import (
	"context"
	"time"

	"github.com/shopspring/decimal"

	"github.com/waqasraza123/trading-data-analysis-agent/apps/go/market-worker/internal/candles"
)

type MockProvider struct{}

func (MockProvider) Key() string {
	return "mock_polling"
}

func (MockProvider) FetchCandles(ctx context.Context, request FetchCandlesRequest) (FetchCandlesResult, error) {
	timeframe, err := candles.ParseTimeframe(request.Timeframe)
	if err != nil {
		return FetchCandlesResult{}, NewProviderError("unsupported_timeframe", "Unsupported timeframe")
	}
	start := time.Date(2026, 4, 29, 10, 0, 0, 0, time.UTC)
	if request.StartTime != nil {
		start = request.StartTime.UTC()
	}
	limit := request.Limit
	if limit <= 0 {
		limit = 100
	}
	items := make([]candles.ProviderCandleInput, 0, limit)
	for index := 0; index < limit; index++ {
		select {
		case <-ctx.Done():
			return FetchCandlesResult{}, ctx.Err()
		default:
		}
		timestamp := start.Add(time.Duration(index) * timeframe.Duration())
		if request.EndTime != nil && timestamp.After(request.EndTime.UTC()) {
			break
		}
		openValue := decimal.NewFromInt(65000 + int64(index))
		closeValue := openValue.Add(decimal.NewFromInt(5))
		highValue := closeValue.Add(decimal.NewFromInt(10))
		lowValue := openValue.Sub(decimal.NewFromInt(10))
		volumeValue := decimal.NewFromInt(10 + int64(index))
		volumeText := volumeValue.String()
		items = append(items, candles.ProviderCandleInput{
			Timestamp: timestamp,
			Open:      openValue.StringFixed(2),
			High:      highValue.StringFixed(2),
			Low:       lowValue.StringFixed(2),
			Close:     closeValue.StringFixed(2),
			Volume:    &volumeText,
			IsFinal:   true,
			ProviderMetadata: map[string]any{
				"provider":       "mock_polling",
				"providerSymbol": request.ProviderSymbol,
				"sequence":       index + 1,
			},
			RawItem: map[string]any{
				"provider":       "mock_polling",
				"sequence":       index + 1,
				"providerSymbol": request.ProviderSymbol,
			},
		})
	}
	return FetchCandlesResult{
		Candles: items,
		ProviderMetadata: map[string]any{
			"provider":  "mock_polling",
			"generated": true,
		},
	}, nil
}
